### ISSUE TITLE
Key Vault Access Policies Not Being Followed

### ISSUE THREAD

**Issue by uipo78** [Azure/go-autorest#477]
(It's a bit hard for me to create a concise description of the source and symptom of the error that I'm seeing since I'm not a contributor to the repository that I'm going to reference, so bear with me :D)

[This issue on Mozilla's SOPS CLI repository](https://github.com/mozilla/sops/issues/537) describes a bug whose source may reside somewhere within this repository. 

Summary of the problem&mdash;
`sops` authenticates a user using this SDK and attempts to use their credentials in order to (in this case) gain access to use a key in an Azure Key Vault for encryption. However, Azure says the operation `encrypt` is not permitted, despite the fact that a policy exists in the Key Vault which allows the authenticated user to use that key for encryption.

Let me know if this problem should be addressed elsewhere.

Thanks!

**Comment by jhendrixMSFT** [comment]
You are sure that the OID of the user (sounds like you're not using a service principal) matches the OID in the policy?  If you run `az keyvault show -n {vault-name}` it's the same OID in the list?

**Comment by uipo78** [comment]
I just verified that the OID emitted in `sops`'s logs matches my user OID. However, the access policy on my Key Vault corresponds to an Azure AD group (of which I am a member), meaning that the OID emitted by `sops` doesn't match the OID listed by `az keyvault show -n {vault-name}`. 

Heyyyy, therein lies the problem if the user's OID has to match an OID listed by `az keyvault show -n {vault-name}`.

**Comment by jhendrixMSFT** [comment]
Hmm per the [docs](https://docs.microsoft.com/en-us/azure/key-vault/key-vault-secure-your-key-vault#data-plane-and-access-policies) this is supposed to work.

I looked at the error message from the linked bug, I see the following.
```
appid=<redacted>
```
Do you know what OID this is referring to, i.e. is it the authenticated user or a service principal?  Is it part of the requisite group?

**Comment by uipo78** [comment]
I have no idea which OID that appid is referring to. Nothing showed up when I searched for it in Azure AD either.

**Comment by uipo78** [comment]
I suspect that it's the same user as we were talking about above, considering the `Caller:` part preceding it.

**Comment by uipo78** [comment]
the part in question of that error (which im guessing is coming from this repo's code) would be displayed more neatly on a Windows system like so
```
Operation: "encrypt"
Caller: appid=<redacted>;oid=<redacted>;numgroups=<redacted>;iss=<redacted>
Vault: <redacted>;location=<redacted>
```

**Comment by uipo78** [comment]
Is there any other information that I can provide to help you out?

**Comment by jhendrixMSFT** [comment]
Sorry for the delay.

Just to confirm, in the redacted output above, the `oid` returned is in the AAD group that has encrypt permissions but the `appid` is not?  I'm not sure how the two relate, I'll find an answer and post an update.

**Comment by jhendrixMSFT** [comment]
Spoke with a guy that works on key vault, the `appid` listed in the error message is what should be in the AAD group that has encrypt access.

How are you creating the authorizer for the key vault client?

**Comment by uipo78** [comment]
Aha, I can confirm that the `appid` in the error message does not correspond to the AAD group with encrypt access.

I'm not a contributor on the `sops` repo, but [this](https://github.com/mozilla/sops/blob/2544b3cdb032ca34e010799bd3fe795b8477a6d5/azkv/keysource.go#L142) appears to be the section where they create an authorizer given a CLI-based login scheme (that's what I attempted to do in the original issue).

**Comment by jhendrixMSFT** [comment]
When you say "`appid` in the error message does not correspond to the ADD group with encrypt access" does that mean that the `appid` isn't in the group?

**Comment by uipo78** [comment]
Whoops I misread your comment. I thought you were saying that the `appid` should correspond to the `appid` of the AAD group that has encrypt access.

`appid` isn't in the group. I still have no idea what it corresponds to either.  Nothing showed up when I searched for it in the Azure AD portal.

**Comment by uipo78** [comment]
A few things that I don't understand:
* Where does this `appid` come from?
* My user (and so, by extension, my `oid`) is in the AAD group with encrypt access. Where does this `appid` come into play if I can find my membership to the AAD group with encrypt access from my `oid`?

**Comment by jhendrixMSFT** [comment]
I set up a repro for this and I think the `appid` isn't relevant.  Sorry for the confusion.

In the portal I created a test group with no members (I was the owner).  I then created an access policy in my key vault using the name of the group I just created with only `decrypt` key permissions.  I then ran my test app which uses `auth.NewAuthorizerFromCLIWithResource()` to authenticate as the currently signed in CLI user (me).  I received an `AccessDenied` as expected since my group contained no members.

Next, I added myself to the test group (via the portal) using the `oid` returned in the error message.  I confirmed it matched the `objectId` in the output from `az ad signed-in-user show`.  I then ran `az ad group member list --group <groupID>` to confirm that my `oid` was in the list (it was).  I ran the repro again and received a `ForbiddenByPolicy` error which is expected since I only have `decrypt` permissions set.

Finally, I added the `encrypt` permission for my test group via the portal, ran the repro and it worked.  I ran `az keyvault show <vault name>` and in the array of access policies I verified that there was an object ID that matched the group I created (i.e. the same <group ID> I used in `az ad group member list` above) and that the group has the `Encrypt` permissions set for that group.
```json
    "accessPolicies": [
      {
        "applicationId": null,
        "objectId": "<group ID>",
        "permissions": {
          "certificates": [],
          "keys": [
            "Decrypt",
            "Encrypt"
          ],
          "secrets": [],
          "storage": null
        },
        "tenantId": "<tenant ID>"
      }
    ],
```
